R/methods-MbioDataset.R

```diff
@@ -283,4 +283,119 @@ setMethod("getCollection", "MbioDataset", function(object, collectionName = char
     } 
 
     return(abundanceData)
+})
+
+collectionVarNamesGeneric <- getGeneric("getCollectionVariableNames", "veupathUtils")
+#' Get Microbiome Dataset Collection Variable Names
+#' 
+#' Get the variable names in a collection in the Microbiome Dataset.
+#' 
+#' @examples
+#' variableNames <- getCollectionVariableNames(microbiomeData::DiabImmune, "16S (V4) Genus")
+#' @param object A Microbiome Dataset
+#' @param collectionName The name of the collection to return the variable names for
+#' @return a character vector of the variable names in the requested collection
+#' @export
+#' @rdname getCollectionVariableNames
+#' @aliases getCollectionVariableNames,MbioDataset,character-method
+setMethod(collectionVarNamesGeneric, "MbioDataset", function(object, collectionName) {
+    return(veupathUtils::getCollectionVariableNames(getCollection(object, collectionName)))
+})
+
+#' Get Microbiome Dataset Variables
+#' 
+#' Get the variables in the Microbiome Dataset by their names.
+#' The requested variables could belong to any collection or
+#' to the metadata. The returned data.table will contain the 
+#' requested variables as columns. If one of the requested
+#' variables cannot be returned, a warning will be printed.
+#' 
+#' @examples
+#' getCollectionVariableNames(microbiomeData::DiabImmune, "16S (V4) Genus")
+#' getMetadataVariableNames(microbiomeData::DiabImmune)
+#' variablesDT <- getVariables(
+#'      microbiomeData::DiabImmune, 
+#'      list("metadata" = c("age_months", "sex"),
+#'           "16S (V4) Genus" = "Bacteroides", 
+#'           "WGS Metagenome enzyme pathway abundance data" = "ANAGLYCOLYSIS-PWY: glycolysis III (from glucose)"
+#'      )
+#' )
+#' @param object A Microbiome Dataset
+#' @param variables The names of the variables to return. this should be a named list
+#' where the names are collection names and the values are variable names for that collection.
+#' For the case of metadata variables, the name should be "metadata".
+#' @return a data.table of the requested variables
+#' @rdname getVariables
+#' @export
+setGeneric("getVariables", function(object, variables) standardGeneric("getVariables"), signature = "object")
+
+#' @rdname getVariables
+#' @aliases getVariables,MbioDataset,character-method
+setMethod("getVariables", "MbioDataset", function(object, variables) {
+
+    if (!is.list(variables)) {
+        stop("variables argument must be a list")
+    }
+    if (is.null(names(variables))) {
+        stop("variables argument must be a named list")
+    }
+
+    ## identify variables w identical names early
+    flattenedVars <- unlist(variables)
+    dups <- unname(flattenedVars[duplicated(flattenedVars)])
+    collectionsWithDups <- lapply(variables, function(x) {dups %in% x})
+    collectionsWithDupsIndexes <- unname(which(collectionsWithDups == TRUE))
+
+    fetchCollectionVariables <- function(collectionIndex) {
```

Comment:
Just a suggestion. I understand why to have the function defined here, but i personally find it hard to follow when functions are defined in functions. If reasonable, i'd move the def outside of a function. But that is just my preference.

Reply:
i tend to like them nested if i kind of actively dont want someone trying to reuse it for random things.. if i move it ill have to generalize it in a way that it could actually be properly repurpose-able. im not sure how useful or worth the time that is..